### PR TITLE
Port Spinner from Gatsby to Next.js(TS)

### LINKED ISSUE #1438: Port SearchPage from Gatsby to NextJS
**What would you like to be added**:
This is an issue to port SearchPage from Gatsby to NextJS

**Why would you like this to be added**:
Part of #1316

### LINKED ISSUE #1479: Port Spinner from Gatsby to Next.js(TS)
What would you like to be added:
This is an issue to port Spinner from Gatsby to NextJS

Why would you like this to be added:
Part of #1316

### PR DESCRIPTION
**Pull request by c3ho**
## Issue This PR Addresses
Fixes #1479 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
Porting from Gatsby to Next JS + changing JS to TS for the `Spinner` component.

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/frontend/next/src/components/SearchPage/SearchPage.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import React, { FC, FormEvent, useState } from 'react';
```
Comment (manekenpix) [inline src/frontend/next/src/components/SearchPage/SearchPage.tsx:1]: In [this](https://github.com/vercel/next.js/issues/12964#issuecomment-629642157) thread:

> Next.js automatically adds the React import when JSX is used indeed. However keep in mind that we do still need to import React from 'react' when the React variable is used.

File: src/frontend/next/src/components/SearchPage/SearchResults.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { Container, Box, Grid, Typography, ListSubheader } from '@material-ui/core';
+
+import useSiteMetadata from '../../hooks/use-site-metadata';
+import Timeline from '../Posts/Timeline.jsx';
+import Spinner from '../Spinner';
+
+type SearchPageProps = {
+  text: string;
+  filter: string;
+};
```
Comment (manekenpix) [inline src/frontend/next/src/components/SearchPage/SearchResults.tsx:13]:
```suggestion
interface SearchPageProps {
  text: string | null;
  filter: string | null;
};
```

**Comment by rogercyyu** [comment]
whoops, sorry. Accidentally clicked that button.

**Comment by humphd** [comment]
@c3ho could you rebase this for package.json changes?

File: src/frontend/next/package.json [hunk 3]
```diff
@@ -8,9 +8,13 @@
...
   },
   "dependencies": {
+    "@material-ui/core": "^4.11.1",
     "next": "^10.0.2",
+    "query-string": "^6.13.7",
     "react": "^16.13.1",
-    "react-dom": "^16.13.1"
+    "react-dom": "^16.13.1",
+    "swr": "^0.3.9",
+    "use-query-params": "^1.1.9"
```
Comment (chrispinkney) [inline src/frontend/next/package.json:17]: All of these are already merged into next now

File: src/frontend/next/src/components/Spinner/Spinner.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+const useStyles = makeStyles((theme) => ({
+  root: {
+    display: 'flex',
+    '& > * + *': {
+      marginLeft: theme.spacing(2),
+    },
+  },
+}));
+
+const Spinner: FC = () => {
```
Comment (chrispinkney) [inline src/frontend/next/src/components/Spinner/Spinner.tsx:14]:
```suggestion
import { makeStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    '& > * + *': {
      marginLeft: theme.spacing(2),
    },
  },
}));

const Spinner = () => {
```

AFAIK we don't need to import `FC`

File: src/frontend/next/src/components/SearchPage/SearchPage.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import { FC, FormEvent, useState } from 'react';
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchPage.tsx:1]:
```suggestion
import { FormEvent, useState } from 'react';
```

File: src/frontend/next/src/components/SearchPage/SearchPage.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import BackToTopButton from '../BackToTopButton';
+
+const useStyles = makeStyles(() => ({
+  anchor: {
+    position: 'absolute',
+    top: 0,
+  },
+}));
+
+const SearchPage: FC = () => {
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchPage.tsx:16]:
```suggestion
const SearchPage = () => {
```

File: src/frontend/next/src/components/SearchPage/SearchPage.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  const classes = useStyles();
+  // We synchronize the `text` and `filter` values to the URL's query string
+  // via `textParam` and `filterParam`. The <SearchBar> UI uses our internal
+  // state values, and the <SearchResults> only update on page load or when
+  // the user submits the form.
+  const [textParam = '', setTextParam] = useQueryParam('text', StringParam);
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchPage.tsx:22]: So I merged in the current master to this pr and I'm getting errors when trying to render ` <SearchPage />`:

`Error: useQueryParams must be used within a QueryParamProvider`

Got any ideas?

File: src/frontend/next/src/components/SearchPage/SearchResults.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
+import { FC } from 'react';
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchResults.tsx:1]:
```suggestion

```

File: src/frontend/next/src/components/SearchPage/SearchResults.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    justifyContent: 'center',
+  },
+  searchResults: {
+    padding: 0,
+    width: '100%',
+    justifyContent: 'center',
+  },
+}));
+
+const SearchResults: FC<SearchPageProps> = ({ text, filter }: SearchPageProps) => {
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchResults.tsx:27]:
```suggestion
const SearchResults = ({ text, filter }: SearchPageProps) => {
```

File: src/frontend/next/src/components/SearchPage/SearchResults.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      const res = await fetch(u);
+      const results = await res.json();
+      return results.values;
+    }
+  );
+
+  if (error) {
+    return (
+      <Container className={classes.searchResults}>
+        <Box className={classes.root} boxShadow={2} marginTop={10}>
```
Comment (chrispinkney) [inline src/frontend/next/src/components/SearchPage/SearchResults.tsx:50]: Getting `Property 'root' does not exist on type 'Record<"spinner" | "searchResults", string>'.` with the latest master merge, same with line 52 and 53.

But I'm getting `No search results` when rendering the component so I think it's fine.

**Comment by chrispinkney** [comment]
So that was a lot, to the point where I feel like I might be doing something wrong on my end, haha. Let me know what you think, plz n thanks!

**Review by manekenpix — APPROVED** [review]
LGTM :+1: 

Peek 2021-01-20 10-17

**Review by chrispinkney — APPROVED** [review]
😍

**Pull request merged**